### ISSUE TITLE
[BUG]: Command 'Copyright: Insert Header' resulted in an error (Unexpected token } in JSON at position 1182)

### ISSUE THREAD

**Issue by max-vogler** [minherz/copyright-inserter#17]
### Plugin Version

v1.0.11

### What happened?

An error popup is shown when trying to insert the copyright header into a TypeScript file

```
Command 'Copyright: Insert Header' resulted in an error (Unexpected token } in JSON at position 1182)
```

### What did you expect to happened?

I expected the copyright header to be inserted into a `.ts` file. Inserting into TypeScript files stopped working after a restart, inserting into HTML still works.

### Relevant log output

```shell
[error] SyntaxError: Unexpected token } in JSON at position 1182
	at JSON.parse (<anonymous>)
	at CopyrightInserter.loadLanguageConfiguration (~/.vscode/extensions/minherz.copyright-inserter-1.0.11/out/copyrightinserter.js:205:21)
	at CopyrightInserter.getLanguageConfigById (~/.vscode/extensions/minherz.copyright-inserter-1.0.11/out/copyrightinserter.js:159:40)
	at CopyrightInserter.insertHeader (~/.vscode/extensions/minherz.copyright-inserter-1.0.11/out/copyrightinserter.js:109:29)
	at insertHeader (~/.vscode/extensions/minherz.copyright-inserter-1.0.11/out/extension.js:32:9)
	at f._executeContributedCommand (~/Downloads/Visual Studio Code.app/Contents/Resources/app/out/vs/workbench/api/node/extensionHostProcess.js:85:30255)
	at f.$executeContributedCommand (~/Downloads/Visual Studio Code.app/Contents/Resources/app/out/vs/workbench/api/node/extensionHostProcess.js:85:30797)
	at n._doInvokeHandler (~/Downloads/Visual Studio Code.app/Contents/Resources/app/out/vs/workbench/api/node/extensionHostProcess.js:89:13812)
	at n._invokeHandler (~/Downloads/Visual Studio Code.app/Contents/Resources/app/out/vs/workbench/api/node/extensionHostProcess.js:89:13496)
	at n._receiveRequest (~/Downloads/Visual Studio Code.app/Contents/Resources/app/out/vs/workbench/api/node/extensionHostProcess.js:89:12158)
	at n._receiveOneMessage (~/Downloads/Visual Studio Code.app/Contents/Resources/app/out/vs/workbench/api/node/extensionHostProcess.js:89:10836)
	at ~/Downloads/Visual Studio Code.app/Contents/Resources/app/out/vs/workbench/api/node/extensionHostProcess.js:89:8945
	at d.invoke (~/Downloads/Visual Studio Code.app/Contents/Resources/app/out/vs/workbench/api/node/extensionHostProcess.js:58:145)
	at w.fire (~/Downloads/Visual Studio Code.app/Contents/Resources/app/out/vs/workbench/api/node/extensionHostProcess.js:58:1856)
	at c.fire (~/Downloads/Visual Studio Code.app/Contents/Resources/app/out/vs/workbench/api/node/extensionHostProcess.js:66:19017)
	at ~/Downloads/Visual Studio Code.app/Contents/Resources/app/out/vs/workbench/api/node/extensionHostProcess.js:104:34267
	at d.invoke (~/Downloads/Visual Studio Code.app/Contents/Resources/app/out/vs/workbench/api/node/extensionHostProcess.js:58:145)
	at w.fire (~/Downloads/Visual Studio Code.app/Contents/Resources/app/out/vs/workbench/api/node/extensionHostProcess.js:58:1856)
	at c.fire (~/Downloads/Visual Studio Code.app/Contents/Resources/app/out/vs/workbench/api/node/extensionHostProcess.js:66:19017)
	at s._receiveMessage (~/Downloads/Visual Studio Code.app/Contents/Resources/app/out/vs/workbench/api/node/extensionHostProcess.js:66:23598)
	at ~/Downloads/Visual Studio Code.app/Contents/Resources/app/out/vs/workbench/api/node/extensionHostProcess.js:66:21132
	at d.invoke (~/Downloads/Visual Studio Code.app/Contents/Resources/app/out/vs/workbench/api/node/extensionHostProcess.js:58:145)
	at w.fire (~/Downloads/Visual Studio Code.app/Contents/Resources/app/out/vs/workbench/api/node/extensionHostProcess.js:58:1856)
	at w.acceptChunk (~/Downloads/Visual Studio Code.app/Contents/Resources/app/out/vs/workbench/api/node/extensionHostProcess.js:66:15848)
	at ~/Downloads/Visual Studio Code.app/Contents/Resources/app/out/vs/workbench/api/node/extensionHostProcess.js:66:14978
	at Socket.P (~/Downloads/Visual Studio Code.app/Contents/Resources/app/out/vs/workbench/api/node/extensionHostProcess.js:104:13790)
	at Socket.emit (events.js:315:20)
	at addChunk (internal/streams/readable.js:309:12)
	at readableAddChunk (internal/streams/readable.js:284:9)
	at Socket.Readable.push (internal/streams/readable.js:223:10)
	at Pipe.onStreamRead (internal/stream_base_commons.js:188:23) copyright.insertHeader {"value":"minherz.copyright-inserter","_lower":"minherz.copyright-inserter"}
```

**Comment by minherz** [comment]
hi,
i am sorry to hear about your experience. i've tested the behavior both on an empty file and the a file with typescript code that VSCode identified as "TypeScript" and failed reproducing the result you are describing.

It will be helpful if you can provide the following information:
- the version of the VS Code you are using
- the Operating System you are running
- the context of the .ts file for which you invoke the extension
- the way you invoke the extension (e.g. shortcut, from command palette, etc.)

Thank you

**Comment by max-vogler** [comment]
I invoked the extension from the command palette:
<img width="738" alt="image" src="https://user-images.githubusercontent.com/864168/158155818-e33f8f79-70cf-4433-ad70-7c17ed582933.png">

The file context is a regular open-source repository: https://github.com/google/pixelate

I found the above log lines in the Output tab of "Log (Extension Host)" and peeked a little into the referenced files but could not quickly find the JSON file in question. If you have any pointers on which JSON file could cause the problem I'll gladly send more debugging data!

My VSCode and operating system version is:
```
Version: 1.65.1
Commit: 8908a9ca0f221f36507231afb39d2d8d1e182702
Date: 2022-03-08T02:20:11.670Z
Electron: 13.5.2
Chromium: 91.0.4472.164
Node.js: 14.16.0
V8: 9.1.269.39-electron.0
OS: Darwin x64 21.3.0
```

**Comment by minherz** [comment]
Thank you for the provided info. I just checked it with [main.ts](https://github.com/google/pixelate/blob/main/src/main.ts) from the provided repository. Without removing already existing copyright there is no changes to the file. If the existing copyright is removed and the plugin is invoked a new copyright is added as expected.
I tested it on VSCode 1.63.2:
```
Version: 1.63.2
Commit: 899d46d82c4c95423fb7e10e68eba52050e30ba3
Date: 2021-12-15T09:39:46.686Z
Electron: 13.5.2
Chromium: 91.0.4472.164
Node.js: 14.16.0
V8: 9.1.269.39-electron.0
OS: Linux x64 5.15.15-1rodete2-amd64
```
Unfortunately, I can check it only on a work laptop which limits me to versions of VSCode that I can use. But the rest is the same.
One more question. Do you use any of the "Remote development" features of the VSCode? Is it possible that you run it in the remote mode?

**Comment by minherz** [comment]
@max-vogler do you still have this issue? if you do, i will need some assistance from you in order to reproduce the problem. please, reply about the work mode. i believe you would not be able to test it vs "my" version of the vscode. maybe you can upgrade to the latest version of the code and see how it works

**Comment by max-vogler** [comment]
Yes, I still encounter this issue. I do have the Remote Development extension pack installed and use it for other projects, but only use copyright inserter with a git repository that is present locally, without any remote features. I'll gladly investigate if you can give me some starters on which files might be culprits.

**Comment by minherz** [comment]
The only way I can think of is to debug the extension. For this you will have to clone the repo and open it in vscode. Then run it and open the file that cause the problem from the vscode debug window and trace/debug the invocation. You can find more instructions in [vscode debug extension](https://code.visualstudio.com/api/extension-guides/debugger-extension) doc.

I have 1.66.0 version now. If you copy/paste or provide a link to gpaste with the full code of the file and the exact file name and extension, i can try to reproduce it with my current version.

**Comment by rolud** [comment]
Hi @minherz, I debugged the extensions and found that the problem were trailing commas in the language configuration json. I created a pr #18

**Comment by minherz** [comment]
Thank you for taking time and investigating it. I will look into PR shortly. Would you mind to provide more details results about your investigation (here or in #18 description)? 
While I tried to reproduce the described behavior I did not see any problem. I used the OoTB TS language packages. Is it possible that you found a problem in a custom language package? In this case, can you please provide a reference to the language package or file that cause the problem?
Before merging this change, I'd like to make sure that this change would not have side effect on other language packages.

Thank you

**Comment by minherz** [comment]
@rolud I have looked into `resources/app/extensions/typescript-basics/language-configuration.js` and its syntax except for using comments is valid.
Can you please elaborate which trailing commas you are referencing?
I'd like to make sure that the introduced change would not compromise other language configurations. Also if the current problem is in the format of the single configuration file, it maybe worth to open an issue to fix the incorrect syntax.

**Comment by rolud** [comment]
The problem is the trailing comma at the end of an object or an array. This one is an example: https://github.com/microsoft/vscode/blob/d9e91c33351c051876c731485a0c47a5852b904b/extensions/typescript-basics/language-configuration.json#L179.

`JSON.parse` throws an error if is present a trailing comma in the json string: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Trailing_commas.

**Comment by minherz** [comment]
I found it, thank you. I think that given it is a second fix that corrects just another out of JSON standard formatting (the first was comment lines), it would be more productive to use the same method for JSON parsing that VSCode itself is using. That method is clearly less restrictive since these configuration files are parsed and processed by VSCode. WDYT?

**Comment by minherz** [comment]
It seems that VSCode uses [jsonc-parser](https://www.npmjs.com/package/jsonc-parser) (v2.2.1). Would you mind to see if it can be used? I am in transfer right now and wait for my shipment including my workstation to arrive.

**Comment by rolud** [comment]
> It seems that VSCode uses [jsonc-parser](https://www.npmjs.com/package/jsonc-parser) (v2.2.1). 

Ok, I will try to use it.

**Comment by rolud** [comment]
I removed checks on comments and trailing commas, and replaced JSON.parse with jsonc-parser. It seems to work, I tested the extension with some languages (typescript, javascript, java, c, python, ruby) with no problems.

**Comment by minherz** [comment]
Thank you very much! The PR is approved. I will release the next version of the plugin during this week.

**Comment by max-vogler** [comment]
Thank you for fixing this issue, @rolud!

**Comment by minherz** [comment]
The fix is released under 1.0.12. The proposed version of the parser was reported having few critical and high vulnerabilities. I ended up to upgrade it to the more secure version. Thank you, @rolud for your contribution.

My apologies that it took so long to release the updated version of the extension.
